File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
import java.util.ArrayDeque;
import java.util.Deque;

public class Solution {

    public static long solution(int cap, int n, int[] deliveries, int[] pickups) {
        long answer = 0;

        Deque<Integer> del = new ArrayDeque<>();
        Deque<Integer> pick = new ArrayDeque<>();

        for (int i = 0; i < n; ++i) {
            if (deliveries[i] > 0) {
                del.push(i);
            }

            if (pickups[i] > 0) {
                pick.push(i);
            }
        }


        while (!del.isEmpty() || !pick.isEmpty()) {

            int delCount = cap;
            int pickCount = cap;

            long max  = 0;

            while(!del.isEmpty() && delCount > 0) {
                int dIdx = del.peek();

                if (dIdx > max)
                    max = dIdx;

                if (delCount >= deliveries[dIdx])
                {
                    delCount -= deliveries[dIdx];
                    del.pop();
                }
                else
                {
                    deliveries[dIdx] -= delCount;
                    break;
                }
            }

            while(!pick.isEmpty() && pickCount > 0) {
                int pIdx = pick.peek();

                if (pIdx > max)
                    max = pIdx;

                if (pickCount >= pickups[pIdx])
                {
                    pickCount -= pickups[pIdx];
                    pick.pop();
                }
                else
                {
                    pickups[pIdx] -= pickCount;
                    break;
                }
            }

            answer += (max + 1) * 2L;
        }

        return answer;
    }
}

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

long long solution(int cap, int n, vector<int> deliveries, vector<int> pickups) {
    long long answer = 0;
    priority_queue<int, vector<int>, less<int>> d;
    priority_queue<int, vector<int>, less<int>> p;

    for (int i = 0; i < n; ++i)
    {
        if (deliveries[i] > 0)
        {
            d.push(i);
        }

        if (pickups[i] > 0)
        {
            p.push(i);
        }
    }

    
    while (!p.empty() || !d.empty())
    {
        int MaxDistance = 0;
        int pCount = cap;
        while (!p.empty())
        {
            int houseIndex = p.top();

            MaxDistance = max(MaxDistance, houseIndex + 1);

            p.pop();

            pCount -= pickups[houseIndex];


            if (pCount <= 0)
            {
                if (pCount < 0)
                {
                    pickups[houseIndex] = abs(pCount);
                    p.push(houseIndex);
                }
                break;
            }
        }

        int dCount = cap;
        while (!d.empty())
        {
            int houseIndex = d.top();
            MaxDistance = max(MaxDistance, houseIndex + 1);
            d.pop();

            dCount -= deliveries[houseIndex];

            if (dCount <= 0)
            {
                if (dCount < 0)
                {
                    deliveries[houseIndex] = abs(dCount);
                    d.push(houseIndex);
                }
                break;
            }
        }

        answer += MaxDistance * 2;
    }


    return answer;
}